util/schema.go

```diff
@@ -0,0 +1,326 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/openconfig/goyang/pkg/yang"
+)
+
+// IsLeafRef reports whether schema is a leafref schema node type.
+func IsLeafRef(schema *yang.Entry) bool {
+	if schema == nil || schema.Type == nil {
+		return false
+	}
+	return schema.Type.Kind == yang.Yleafref
+}
+
+// IsChoiceOrCase returns true if the entry is either a 'case' or a 'choice'
+// node within the schema. These are schema nodes only, and the code generation
+// operates on data tree paths.
+func IsChoiceOrCase(e *yang.Entry) bool {
+	if e == nil {
+		return false
+	}
+	return e.IsChoice() || e.IsCase()
+}
+
+// SchemaPaths returns all the paths in the path tag.
+func SchemaPaths(f reflect.StructField) ([][]string, error) {
+	var out [][]string
+	pathTag, ok := f.Tag.Lookup("path")
+	if !ok || pathTag == "" {
+		return nil, fmt.Errorf("field %s did not specify a path", f.Name)
+	}
+	if pathTag == "" {
+		return out, nil
+	}
+
+	ps := strings.Split(pathTag, "|")
+	for _, p := range ps {
+		out = append(out, StripModulePrefixes(strings.Split(p, "/")))
+	}
+	return out, nil
+}
+
+// ChildSchema returns the first child schema that matches path from the given
+// schema root. When comparing the path, only nodes that appear in the data tree
+// are considered. It returns nil if no node matches the path.
+func ChildSchema(schema *yang.Entry, path []string) *yang.Entry {
+	path = StripModulePrefixes(path)
+	entries := FindFirstNonChoiceOrCase(schema)
+
+	for _, e := range entries {
+		m := MatchingNonChoiceCaseSchema(e, path)
+		if m != nil {
+			return m
+		}
+	}
+
+	return nil
+}
+
+// FindFirstNonChoiceOrCase recursively traverses the schema tree and returns a
```

Comment:
Ditto, now this is public, we have multiple implementations of this function (the other is in `ygen`), let's create an issue/TODO to fix that.

Reply:
Grouped in comment with the same issue since these are related.